src/ssh/cockpitsshrelay.c

```diff
@@ -1320,61 +1320,67 @@ send_auth_reply (CockpitSshData *data,
   return ret;
 }
 
-static void
+static gboolean
 parse_host (const gchar *host,
             gchar **hostname,
             gchar **username,
             guint *port)
 {
-  gchar *user_arg = NULL;
-  gchar *host_arg = NULL;
-  gchar *tmp = NULL;
-  gchar *end = NULL;
-
-  guint64 tmp_num;
-
-  gsize host_offset = 0;
-  gsize host_length = strlen (host);
-
-  tmp = strrchr (host, '@');
-  if (tmp)
-    {
-      if (tmp[0] != host[0])
+  GError *error = NULL;
+  g_autoptr (GRegex) regex = g_regex_new ("^"
+         "(?:(.+)@)?"         /* optional username */
+         "(?|"                /* one of... */
+           "\\[([^]@]+)\\]"     /* hostname in square brackets, no @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@:]+)"           /* hostname with no : or @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@]+)"            /* hostname with no @ but : (IPv6 address), and no port */
+         ")"                  /* . */
+       "$",
+       0, 0, &error);
+  g_assert_no_error (error);
+
+  g_autoptr(GMatchInfo) info = NULL;
+
+  if (g_regex_match (regex, host, 0, &info))
+    {
+      g_autofree gchar *port_str = g_match_info_fetch (info, 3);
+
+      if (port_str && port_str[0] != '\0')
         {
-          user_arg = g_strndup (host, tmp - host);
-          host_offset = strlen (user_arg) + 1;
-          host_length = host_length - host_offset;
+          gchar *endptr;
+          gint64 value = g_ascii_strtoull (port_str, &endptr, 10);
```

Comment:
Careful.  That returns a `guint64` and you're implicitly casting it to signed here.  You could well end up with a negative value (and the check below only checks for `value` being too large).

Also: I think it makes sense to reject 0.  0 has the special meaning here of "not specified".  It should not be possible to explicitly specify to connect to port 0 (which is both impossible, and would be interpreted as "default port" anyway).

Reply:
We could tweak the regexp to require the pattern `[1-9][0-9]*` for port number... and then we could replace the entire thing with some code like:

```
  g_autofree gchar *port_str = get_match();
  int p = atoi (port_str ?: "");
  if (0 <= p && p < 65536)
    *port = p;
  else
    complain();
```

Reply:
Argh, well spotted -- that was a typo, I deliberately chose strto*u*ll. I fixed it locally. I still want to look at the [brittle integration test](https://logs.cockpit-project.org/logs/pull-14615-20200915-180548-fbde9afb-debian-testing/log.html#292) -- that's an old root cause, but it bites us again here.

Reply:
Nice idea! Seems to work fine locally. I pushed that, so that the tests can run again. I still want to track down that flake, though.